sphinx_gallery/gen_rst.py

```diff
@@ -415,16 +417,22 @@ def patch_warnings():
 class _exec_once(object):
     """Deal with memory_usage calling functions more than once (argh)."""
 
-    def __init__(self, code, globals_):
+    def __init__(self, code, fake_main):
         self.code = code
-        self.globals = globals_
+        self.fake_main = fake_main
         self.run = False
 
     def __call__(self):
         if not self.run:
             self.run = True
+            old_main = sys.modules.get('__main__', None)
             with patch_warnings():
-                exec(self.code, self.globals)
+                sys.modules['__main__'] = self.fake_main
+                try:
+                        exec(self.code, self.fake_main.__dict__)
+                finally:
+                    if old_main is not None:
+                        sys.modules['__main__'] = old_main
```

Comment:
What is the purpose of using just `try` and `finally`?

Reply:
It is equivalent to:
```
try:
    ...
except:
    raise
finally:
    ...
```
i.e., it always does what is in the `finally` clause regardless of the success or failure of the `try`. In this case we always want to restore the `sys.modules['__main__']` regardless of whether or not our `exec` worked.

Reply:
Ah yes of course, thanks!